### ISSUE TITLE
git worktree and projects

### ISSUE THREAD

**Issue by ulises** [emacs-lsp/lsp-java#247]
This is more of a question rather than a bug (it might be a bug though).

I have noticed that if I open a class file in a `git worktree` lsp (java/mode) gets confused and constantly reports that the file is not part of the project and that only syntax errors will be shown.

When looking at the lsp capabilities I see the same file opened in 2 projects: the git worktree and the main repo.

Perhaps this is working as intended and worktrees simply aren't supported?

**Comment by yyoncho** [comment]
My guess is that git worktree is not recognized as a project by JDTLS 

> I see the same file opened in 2 projects: 

This means that one server connection is used to manage 2 workspace roots and this connection is managing that buffer.

**Comment by ulises** [comment]
> This means that one server connection is used to manage 2 workspace roots and this connection is managing that buffer.

So this is not it then?

> My guess is that git worktree is not recognized as a project by JDTLS

I wonder why though? IIUC a worktree shouldn't be distinguishable from a regular dir?

Thanks for the quick answer btw

**Comment by yyoncho** [comment]
Generally, when you import project root jdtls expects to find pom.xml(or equivalent) in order to build project deps and so on. If you go and open standalone file without the corresponding project infrastructure you will get that error.

**Comment by ulises** [comment]
> you go and open standalone file without the corresponding project infrastructure you will get that error

Right. Apologies for the incomplete info: when I say "I open a file" I mean the file is within a directory structure that has a pom.xml, etc. In fact I get asked if I want to import `path/to/my/worktree` to which I say yes.

**Comment by yyoncho** [comment]
> In fact I get asked if I want to import path/to/my/worktree to which I say yes.

That is fine - the issue is that jdtls does not like that project for some reason. This is not related to using git worktree.

**Comment by ulises** [comment]
> the issue is that jdtls does not like that project for some reason

Right, fair enough. What's puzzling is that things work just fine if I work on the main clone (even when I switch git branches).

My abilities to debug this are limited, so if you have any hints/suggestions I can give it a go. Alternatively I'll carry on not using worktrees (it was more of a nice to have).

**Comment by yyoncho** [comment]
I am not familiar with how worktrees work. You should make sure that the project is buildable from the command line. Also, are you opening the same project twice(same groupId, same projectId)? If yes, that might be causing the issue as well. But generally, if you are able to build from command line and jdtls is not able to import the project that indicates jdtls bug.

**Comment by ulises** [comment]
> You should make sure that the project is buildable from the command line. 

Right, it's buildable from the cmd line.

> Also, are you opening the same project twice(same groupId, same projectId)? If yes, that might be causing the issue as well.

This might be it. I wonder how this is dealt with when one does a git checkout though? After all, IIUC worktrees are similar to checkouts except they can co-exist in parallel and have their own directory structure.

**Comment by yyoncho** [comment]
As I mentioned the issue is not related to git worktree since JDTLS/lsp-java does not care about that. They care about having a buildable project. The server does not support importing the same project twice. The same is true for eclipse I think. The solution would be to remove the original project from the workspace via M-x lsp-workspace-folder-remove and then import the other one or spawn 2 emacs sessions but with different lsp-session-file.

**Comment by ulises** [comment]
>The server does not support importing the same project twice

This must be it then.

Thanks for all your help!
